OrbitBase/ThreadPool.cpp

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "OrbitBase/ThreadPool.h"
+
+#include <list>
+#include <thread>
+
+#include "OrbitBase/Logging.h"
+#include "absl/synchronization/mutex.h"
+
+namespace {
+
+class ThreadPoolImpl : public ThreadPool {
+ public:
+  explicit ThreadPoolImpl(size_t thread_pool_min_size, size_t queue_threshold);
+
+  void Schedule(std::unique_ptr<Action> action) override;
+  void Shutdown() override;
+  void Wait() override;
+
+ private:
+  bool ActionsAvailableOrShutdownInitiated();
+  // Blocking call - returns nullptr if shutdown initiated.
+  std::unique_ptr<Action> PopAction();
+  void CreateWorker();
+  void WorkerFunction();
+
+  absl::Mutex mutex_;
+  std::list<std::unique_ptr<Action>> scheduled_actions_;
+  std::vector<std::thread> worker_threads_;
+  size_t queue_threshold_;
+  bool shutdown_initiated_;
+};
+
+ThreadPoolImpl::ThreadPoolImpl(size_t thread_pool_min_size,
+                               size_t queue_threshold)
+    : queue_threshold_(queue_threshold), shutdown_initiated_(false) {
+  CHECK(thread_pool_min_size > 0);
+  CHECK(queue_threshold > 0);
+
+  absl::MutexLock lock(&mutex_);
+  for (size_t i = 0; i < thread_pool_min_size; ++i) {
+    CreateWorker();
+  }
+}
+
+void ThreadPoolImpl::CreateWorker() {
+  worker_threads_.emplace_back([this] { WorkerFunction(); });
+}
+
+void ThreadPoolImpl::Schedule(std::unique_ptr<Action> action) {
+  absl::MutexLock lock(&mutex_);
+  CHECK(!shutdown_initiated_);
+  scheduled_actions_.push_back(std::move(action));
+  if (scheduled_actions_.size() >= queue_threshold_) {
```

Comment:
In the ThreadPool header, we see "This class implements a simple fixed size thread-pool", however, it seems that the size of the thread pool is dynamic.  Furthermore, it looks like there is no maximum number of threads, this seems a bit dangerous IMO as someone could schedule many actions at once which would in turn spawn many worker threads.  It seems to me that we'd want to set the number of threads in the pool as a proportion of the number of available cores, once it is set, I wouldn't create workers on the fly.  Am I missing something?

Reply:
I changed my mind in the process of implementing. Fixed the comment.

Number of threads does not always correlate with number of cores. For example threads that mostly bound to IO do not consume processor time. These are threads waiting for grpc calls to be executed for example.

There is a danger of creating too many threads. I added another parameter which limits maximum number of threads.